src/iceberg/catalog/memory_catalog.cc

```diff
@@ -0,0 +1,248 @@
+}
+
+bool MemoryCatalog::DropTable(const TableIdentifier& identifier, bool purge) {
+  std::unique_lock lock(mutex_);
+  // TODO(Guotao): Delete all metadata files if purge is true.
+  return root_container_->UnregisterTable(identifier);
+}
+
+Result<std::shared_ptr<Table>> MemoryCatalog::LoadTable(
+    const TableIdentifier& identifier) const {
+  throw IcebergError("not implemented");
+}
+
+Result<std::shared_ptr<Table>> MemoryCatalog::RegisterTable(
+    const TableIdentifier& identifier, const std::string& metadata_file_location) {
+  std::unique_lock lock(mutex_);
+  if (!root_container_->NamespaceExists(identifier.ns)) {
+    return unexpected<Error>({.kind = ErrorKind::kNoSuchNamespace,
+                              .message = "table namespace does not exist"});
+  }
+  if (!root_container_->RegisterTable(identifier, metadata_file_location)) {
+    return unexpected<Error>(
+        {.kind = ErrorKind::kUnknownError, .message = "The registry failed."});
+  }
+  return LoadTable(identifier);
+}
+
+std::unique_ptr<TableBuilder> MemoryCatalog::BuildTable(const TableIdentifier& identifier,
+                                                        const Schema& schema) const {
+  throw IcebergError("not implemented");
+}
+
+/// Implementation of NamespaceContainer
+NamespaceContainer* NamespaceContainer::GetNamespaceContainer(
+    NamespaceContainer* root, const Namespace& namespace_ident) {
+  return GetNamespaceContainerImpl(root, namespace_ident);
+}
+
+const NamespaceContainer* NamespaceContainer::GetNamespaceContainer(
+    const NamespaceContainer* root, const Namespace& namespace_ident) {
+  return GetNamespaceContainerImpl(root, namespace_ident);
+}
+
+bool NamespaceContainer::NamespaceExists(const Namespace& namespace_ident) const {
+  return GetNamespaceContainer(this, namespace_ident) != nullptr;
+}
+
+std::vector<std::string> NamespaceContainer::ListChildrenNamespaces(
+    const std::optional<Namespace>& parent_namespace_ident) const {
+  auto container = this;
+  if (parent_namespace_ident.has_value()) {
+    container = GetNamespaceContainer(this, *parent_namespace_ident);
+    if (!container) return {};
+  }
+
+  std::vector<std::string> names;
+  auto const& children = container->children_;
+  names.reserve(children.size());
+  std::ranges::transform(children, std::back_inserter(names),
+                         [](const auto& pair) { return pair.first; });
+  return names;
+}
+
+bool NamespaceContainer::CreateNamespace(
+    const Namespace& namespace_ident,
+    const std::unordered_map<std::string, std::string>& properties) {
+  auto container = this;
+  bool newly_created = false;
+
+  for (const auto& part_level : namespace_ident.levels) {
+    if (auto it = container->children_.find(part_level);
+        it == container->children_.end()) {
+      container = &container->children_[part_level];
+      newly_created = true;
+    } else {
+      container = &it->second;
+    }
+  }
+
+  if (!newly_created) return false;
```

Comment:
```suggestion
  if (!newly_created) {
    return false;
  }
```

I'd suggest to add curly braces for better readability. Same for below.